controllers/clustertopology_controller_compute.go

```diff
@@ -140,9 +141,129 @@ func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clust
 	return class, nil
 }
 
-// getTemplate gets the object referenced in ref.
+// getCurrentState gets information about the current state of a Cluster by inspecting the state of the InfrastructureCluster,
+// the ControlPlane, and the MachineDeployments associated with the cluster.
+func (r *ClusterTopologyReconciler) getCurrentState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (*clusterTopologyState, error) {
+	var clusterState clusterTopologyState
+
+	// Reference to the InfrastructureCluster can be nil and is expected to be on the first reconcile loop
+	// In this case the method should still be allowed to continue.
+	if cluster.Spec.InfrastructureRef != nil {
+		infra, err := r.getCurrentInfrastructureClusterState(ctx, cluster)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to retrieve Cluster state")
+		}
+		clusterState.infrastructureCluster = infra
+	}
+
+	// Reference to the ControlPlane can be nil, and is expected to be on the first reconcile. In this case the method
+	// should still be allowed to continue.
+	if cluster.Spec.ControlPlaneRef != nil {
+		cp, err := r.getCurrentControlPlaneState(ctx, cluster, class)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to retrieve Cluster state")
+		}
+		clusterState.controlPlane = cp
+	}
+
+	// A cluster may have zero or more MachineDeployments and a Cluster is expected to have zero MachineDeployments on
+	// first reconcile.
+	m, err := r.getCurrentMachineDeploymentState(ctx, cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve Cluster state")
+	}
+	clusterState.machineDeployments = m
+	clusterState.cluster = cluster
+	return &clusterState, nil
+}
+
+// getCurrentInfrastructureClusterState looks for the state of the InfrastructureCluster. If a reference is set but not
+// found, either from an error or the object not being found, an error is thrown.
+func (r *ClusterTopologyReconciler) getCurrentInfrastructureClusterState(ctx context.Context, cluster *clusterv1.Cluster) (*unstructured.Unstructured, error) {
+	infra, err := r.getReference(ctx, cluster.Spec.InfrastructureRef)
+	if err != nil {
+		return nil, errors.Wrap(err, "InfrastructureCluster")
```

Comment:
I'm wondering if this isn't confusing for the reader, given that "InfrastructureCluster" does not adds context to the error, unless if this is read together with the upper level `errors.Wrap` in `getCurrentState`

Personally I would prefer to get a meaningful error message here, and to avoid adding details at the upper level if not necessary, but I'll defer this to @sbueringer given that the original comment was about making this consistent with read cluster class

Reply:
Hm I see multiple options here:
* current option: "failed to retrieve Cluster state" + "InfrastructureCluster" + ""reference is not set"
* verbose option: "failed to retrieve Cluster state" + "failed to get InfrastructureCluster" + ""reference is not set"
* shorter option: "failed to retrieve Cluster state" + "get InfrastructureCluster" + ""reference is not set"
* shortest option: "retrieve Cluster state" + "get InfrastructureCluster" + ""reference is not set"

I personally prefer the last option as it drops all that's not strictly necessary to understand what's going on on the different levels. Prefixes like "failed to" or "error ..." are imho usually not really useful. But I think across our code base we're probably using almost all of them. To be honest I'm not sure what we should do here, apart from that it would be nice to keep it consistent at least in the local file (which means we would have to adjust at least my calls of getReference above).

Maybe we should just discuss once what we regard as the best option and then try to use that one from now on everywhere? Right now I frequently try to look up what we do "locally" and then try to align to that when adding new code to CAPI.

Reply:
/cc @vincepri

Reply:
Maybe we should deal with the logging conventions in a separate issue and then open a PR as the overall topology code matures to improve the logging and make it more consistent - I'll open an issue against the logging convention topic, but I can also do whatever cleanup might be wanted here to unblock this PR.

Reply:
Agree, we should choose one variant for now / this PR and the bigger discussion should be outside of this PR (I would scope it down to error messages not logging)

Reply:
ok to move the discussion on a issue, but for now I'm +1 to return a meaning full error here e.g.

```suggestion
		return nil, errors.Wrap(err, "failed to read %s" , cluster.Spec.InfrastructureRef.kind)
```
And to not wrap in the caller func.

Reply:
The latest version now implements this kind of error reporting - PTAL